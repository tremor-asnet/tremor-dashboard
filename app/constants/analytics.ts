/* Constants for Analytics Page */
export const SALE_STATISTICAL_TYPE = {
  REVENUE: "Revenue",
};
