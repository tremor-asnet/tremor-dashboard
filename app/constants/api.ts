import { ROUTER_API_URL } from "./routes";

export const API_ROUTES = {
  USERS: `${ROUTER_API_URL}/users`,
};
