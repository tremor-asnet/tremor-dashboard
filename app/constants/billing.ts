export const enum AGGREGATION_TYPE {
  SALARY = "Salary",
  PAYPAL = "Paypal",
}

export const enum AGGREGATION_DESCRIPTION {
  SALARY = "Belong Interactive",
  PAYPAL = "Freelance Payment",
}
