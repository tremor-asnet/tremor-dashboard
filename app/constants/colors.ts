type ColorChannelsChartType = {
  [key: number]: string;
  1: string;
  2: string;
  3: string;
  4: string;
};

export const CHANNELS_CHART_COLOR: ColorChannelsChartType = {
  1: "blue-500",
  2: "red-500",
  3: "primary",
  4: "gray-500",
};

export const TRANSACTION_COLOR = {
  INCREASE: "few",
  DECREASE: "attention",
  PENDING: "secondary",
};
