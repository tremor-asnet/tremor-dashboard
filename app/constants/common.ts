export const EXCEPT_KEYS = {
  POSITIVE_INTEGER: ["e", "E", "+", "-", "."],
  POSITIVE_DOUBLE: ["e", "E", "+", "-"],
};

export const DIRECTION = {
  ASC: "asc",
  DESC: "desc",
};
