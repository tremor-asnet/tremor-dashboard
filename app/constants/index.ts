export * from "./routes";
export * from "./regex";
export * from "./messages";
export * from "./profile";
export * from "./commons";
export * from "./flagSrc";
export * from "./images";
export * from "./sections";
export * from "./images";
export * from "./products";
export * from "./colors";
export * from "./units";
export * from "./sales";
export * from "./commons";
export * from "./status";
