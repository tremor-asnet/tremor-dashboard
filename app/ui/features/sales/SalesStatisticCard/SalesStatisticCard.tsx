"use client";

//Libs
import { useState, RefObject } from "react";

//Components
import { Card, Text, Flex, Button } from "@tremor/react";

//Types
import { SalesStatisticData } from "@/types";

// Hooks
import { useOutsideClick } from "@/hooks";

//Constants
import {
  CURRENCY,
  ITEM_ACTION_SALES_DATE,
  SALES_STATISTIC_TYPE,
  UNIT,
  VARIANT_BUTTON,
} from "@/constants";

// Helpers
import { formatAdjustNumber, formattedNumber, moneyFormat } from "@/helpers";

interface SalesStatisticProp {
  statisticsData: SalesStatisticData;
}

const SalesStatisticCard = ({
  statisticsData,
}: SalesStatisticProp): JSX.Element => {
  const { id, type, amount, amountChange, duration, amountChangeType } =
    statisticsData;
  const [isOpenAction, setOpenAction] = useState(false);
  const [currentSalesDate, setCurrentSalesDate] = useState("6 May - 7 May");
  const openActionSalesDate = isOpenAction;

  const salesCardRef = useOutsideClick(() => {
    setOpenAction(false);
  });

  const handleSelectSalesDate = (labelDate: string) => {
    setOpenAction(false);
    setCurrentSalesDate(labelDate);
  };

  const handleToggleAction = (id: string) => {
    setOpenAction(!isOpenAction);
  };

  const formattedAmount =
    {
      [SALES_STATISTIC_TYPE.SALES]: moneyFormat({
        value: amount,
        currency: CURRENCY.DOLLAR,
      }),
      [SALES_STATISTIC_TYPE.CUSTOMERS]: formattedNumber({
        value: amount,
        isDecimalNumber: true,
      }),
      [SALES_STATISTIC_TYPE.AVG_REVENUE]: moneyFormat({
        value: amount,
        currency: CURRENCY.DOLLAR,
      }),
    }[type] || "";

  const formattedTotalAmount =
    {
      [SALES_STATISTIC_TYPE.SALES]: formatAdjustNumber({
        value: amountChange,
        isPositive: amountChangeType,
        unit: UNIT.PERCENT,
      }),
      [SALES_STATISTIC_TYPE.CUSTOMERS]: formatAdjustNumber({
        value: amountChange,
        isPositive: amountChangeType,
        unit: UNIT.PERCENT,
      }),
      [SALES_STATISTIC_TYPE.AVG_REVENUE]: moneyFormat({
        value: amountChange,
        currency: CURRENCY.DOLLAR,
      }),
    }[type] || "";

  const totalAmountColor =
    type === SALES_STATISTIC_TYPE.AVG_REVENUE
      ? "text-secondary dark:!text-secondary"
      : "text-few";

  return (
    <Card
      ref={salesCardRef as RefObject<HTMLDivElement>}
      className="dark:bg-dark-tremor-primary ring-0 max-w-full p-4 2xl:max-w-full border-none relative rounded-xl shadow-md">
      <Flex alignItems="start">
        <Flex className="flex-col w-2/3 md:w-1/2">
          <Flex
            alignItems="start"
            justifyContent="start"
            flexDirection="col"
            data-testid="formatted-amount">
            <Text className="text-md text-secondary dark:text-dark-romance font-semibold tracking-[0.4px]">
              {type}
            </Text>
            <Text className="text-primary dark:text-dark-primary text-xl leading-[33px] font-bold">
              {formattedAmount}
            </Text>
          </Flex>
        </Flex>
        <Flex
          alignItems="end"
          justifyContent="end"
          className="w-1/3 md:w-1/2 cursor-pointer"
          onClick={() => handleToggleAction(id)}>
          <Text className="!text-xs text-secondary dark:text-secondary leading-[21px] tracking-[0.4px]">
            {currentSalesDate}
          </Text>
        </Flex>
        {openActionSalesDate && (
          <div
            data-testid="sales-date-options"
            className="absolute p-2 -right-2 top-8 z-10 bg-white dark:bg-dark-tremor-primary rounded-lg shadow-md">
            {ITEM_ACTION_SALES_DATE.map(item => (
              <Flex key={item.key}>
                <Button
                  className="w-40 justify-start text-tremor-content-title hover:text-tremor-content-title hover:bg-body dark:hover:bg-dark-secondary hover:rounded-md px-4 py-1.5"
                  variant={VARIANT_BUTTON.LIGHT}
                  onClick={() => handleSelectSalesDate(item.label)}>
                  <Text className="font-normal text-sm text-secondary dark:text-lighter hover:text-primary dark:hover:text-white leading-[21px] tracking-[0.13px]">
                    {item.label}
                  </Text>
                </Button>
              </Flex>
            ))}
          </div>
        )}
      </Flex>
      <Flex
        alignItems="start"
        justifyContent="start"
        data-testid="total-amount">
        {amountChange && (
          <Text
            className={`${totalAmountColor} dark:text-few leading-[22px] font-bold`}>
            {formattedTotalAmount}
          </Text>
        )}
        <Text className="ml-1 text-secondary dark:text-dark-romance leading-[21px] tracking-[0.4px]">
          {duration}
        </Text>
      </Flex>
    </Card>
  );
};

export default SalesStatisticCard;
