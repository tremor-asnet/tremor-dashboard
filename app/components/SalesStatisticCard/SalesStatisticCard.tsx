"use client";

//Libs
import { useState } from "react";

//Components
import { Card, Text, Flex, Button } from "@tremor/react";

//Types
import { TSalesStatistic } from "@/types";

//Constants
import { ITEM_ACTION_SALES_DATE } from "@/constants/commons";
import { SALES_STATISTIC_TYPE } from "@/constants/sales";

//Mocks
import { STATISTICS_DATA } from "@/mocks/sales";

// Helpers
import { formatDecimalNumber } from "@/helpers";

interface ISalesStatisticProp {
  statisticsData: TSalesStatistic;
}

const SalesStatisticCard = ({
  statisticsData,
}: ISalesStatisticProp): JSX.Element => {
  const { id, type, amount, totalAmount, duration } = statisticsData;
  const [isOpenAction, setOpenAction] = useState(false);
  const [currentSalesDate, setCurrentSalesDate] = useState("6 May - 7 May");
  const openActionSalesDate = isOpenAction;

  const handleSelectSalesDate = (labelDate: string) => {
    setOpenAction(false);
    setCurrentSalesDate(labelDate);
  };

  const handleToggleAction = (id: string) => {
    setOpenAction(!isOpenAction);
  };

  const formattedAmount =
    {
      [SALES_STATISTIC_TYPE.SALES]: "$230,220",
      [SALES_STATISTIC_TYPE.CUSTOMERS]: formatDecimalNumber(amount),
      [SALES_STATISTIC_TYPE.AVG_REVENUE]: `$${formatDecimalNumber(amount)}`,
    }[type] || "";

  return (
    <Card className="dark:bg-dark-tremor-primary ring-0 max-w-full p-4 lg:max-w-[356px] 2xl:max-w-full border-none relative rounded-xl shadow-md">
      <Flex className="items-start">
        <Flex className="flex-col w-2/3">
          <Flex className="flex-col justify-start items-start">
            <Text className="text-md text-secondary dark:text-dark-romance font-semibold tracking-[0.4px]">
              {type}
            </Text>
            <Text className="text-primary dark:text-dark-primary text-xl leading-[33px] font-bold">
              {formattedAmount}
            </Text>
          </Flex>
          <Flex className="justify-start items-start">
            {totalAmount && (
              <Text className="text-few dark:text-few leading-[22px] font-bold">
                {`+${totalAmount}`}
              </Text>
            )}
            <Text className="ml-1 text-secondary dark:text-dark-romance leading-[21px] tracking-[0.4px]">
              {duration}
            </Text>
          </Flex>
        </Flex>
        <Flex
          className="justify-end items-end w-1/3 cursor-pointer"
          onClick={() => handleToggleAction(id)}>
          <Text className="!text-xs text-secondary dark:text-secondary leading-[21px] tracking-[0.4px]">
            {currentSalesDate}
          </Text>
        </Flex>
        {openActionSalesDate && (
          <div className="absolute p-2 -right-2 top-8 z-10 bg-white rounded-lg shadow-md">
            {ITEM_ACTION_SALES_DATE.map(item => (
              <Flex key={item.key} flex-col>
                <Button
                  className="w-40 justify-start text-tremor-content-title hover:text-tremor-content-title hover:bg-[#f0f2f5] hover:rounded-md px-4 py-1.5"
                  variant="light"
                  onClick={() => handleSelectSalesDate(item.label)}>
                  <Text className="font-normal text-sm text-secondary hover:text-primary leading-[21px] tracking-[0.13px]">
                    {item.label}
                  </Text>
                </Button>
              </Flex>
            ))}
          </div>
        )}
      </Flex>
    </Card>
  );
};

export default SalesStatisticCard;
