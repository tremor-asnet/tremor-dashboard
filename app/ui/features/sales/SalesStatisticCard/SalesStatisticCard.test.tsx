import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";

// Components
import SalesStatisticCard from "./SalesStatisticCard";

// Constants
import { SALES_STATISTIC_TYPE } from "@/constants";

describe("SalesStatisticCard", () => {
  const statisticsData = {
    id: "1",
    type: "Sales",
    amount: 23220,
    amountChange: 213,
    duration: "since last month",
    amountChangeType: 0,
  };

  it("Should match snapshot", () => {
    const component = render(
      <SalesStatisticCard statisticsData={statisticsData} />,
    );
    expect(component).toMatchSnapshot();
  });

  it("calls handleToggleAction function to toggles sales date options", () => {
    render(<SalesStatisticCard statisticsData={statisticsData} />);

    expect(screen.queryByTestId("sales-date-options")).toBeNull();

    // Click on the sales date text to open the options
    fireEvent.click(screen.getByText("6 May - 7 May"));

    expect(screen.queryByTestId("sales-date-options")).toBeTruthy();

    // Click again to close the options
    fireEvent.click(screen.getByText("6 May - 7 May"));

    expect(screen.queryByTestId("sales-date-options")).toBeNull();
  });

  it("calls handleSelectSalesDate function to selects sales date", () => {
    // Mock the useState hook
    const useStateSpy = jest.spyOn(React, "useState");

    // Mock for isOpenAction
    useStateSpy.mockReturnValueOnce([true, jest.fn()]);
    // Mock for currentSalesDate
    useStateSpy.mockReturnValueOnce(["6 May - 7 May", jest.fn()]);

    const { getByTestId } = render(
      <SalesStatisticCard statisticsData={statisticsData} />,
    );

    const button = getByTestId("sales-date-options").querySelector("button")!;

    fireEvent.click(button, { target: { label: "Last 7 days" } });

    expect(useStateSpy).toHaveBeenCalledWith(false);
    expect(useStateSpy).toHaveBeenCalledWith("6 May - 7 May");
  });

  it("renders correct totalAmountColor for AVG_REVENUE type", () => {
    const { getByTestId } = render(
      <SalesStatisticCard
        statisticsData={{
          ...statisticsData,
          type: SALES_STATISTIC_TYPE.AVG_REVENUE,
        }}
      />,
    );

    const totalAmountElement = getByTestId("total-amount");

    expect(
      totalAmountElement.getAttribute("class")?.includes("text-secondary"),
    ).toBe(false);
    expect(
      totalAmountElement
        .getAttribute("class")
        ?.includes("dark:!text-secondary"),
    ).toBe(false);
  });

  test("renders with empty string formattedTotalAmount", () => {
    const { container } = render(
      <SalesStatisticCard
        statisticsData={{
          ...statisticsData,
          type: SALES_STATISTIC_TYPE.AVG_REVENUE,
        }}
      />,
    );

    const totalAmountElementFew = container.querySelector(".text-few");
    expect(totalAmountElementFew).toBeNull();
  });

  it("renders with empty string formattedAmount", () => {
    const statisticsData = {
      id: "1",
      type: "Sales",
      amount: 0,
      amountChange: 0,
      duration: "since last month",
      amountChangeType: 0,
    };

    const { getByTestId } = render(
      <SalesStatisticCard statisticsData={statisticsData} />,
    );

    const totalAmountElement = getByTestId("formatted-amount");

    expect(totalAmountElement.getAttribute("class")?.includes("")).toBe(true);
  });
});
