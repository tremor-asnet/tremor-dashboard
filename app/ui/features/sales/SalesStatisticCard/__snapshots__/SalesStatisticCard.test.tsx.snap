// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SalesStatisticCard Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="tremor-Card-root w-full text-left rounded-tremor-default bg-tremor-background ring-tremor-ring shadow-tremor-card dark:ring-dark-tremor-ring dark:shadow-dark-tremor-card border-tremor-brand dark:border-dark-tremor-brand dark:bg-dark-tremor-primary ring-0 max-w-full p-4 2xl:max-w-full border-none relative rounded-xl shadow-md"
      >
        <div
          class="tremor-Flex-root flex w-full flex-row justify-between items-start"
        >
          <div
            class="tremor-Flex-root flex justify-between items-center flex-col w-2/3 md:w-1/2"
          >
            <div
              class="tremor-Flex-root flex w-full flex-col justify-start items-start"
              data-testid="formatted-amount"
            >
              <p
                class="text-tremor-default text-md text-secondary dark:text-dark-romance font-semibold tracking-[0.4px]"
              >
                Sales
              </p>
              <p
                class="text-tremor-default text-primary dark:text-dark-primary text-xl leading-[33px] font-bold"
              >
                $23,220.00
              </p>
            </div>
          </div>
          <div
            class="tremor-Flex-root flex flex-row justify-end items-end w-1/3 md:w-1/2 cursor-pointer"
          >
            <p
              class="text-tremor-default !text-xs text-secondary dark:text-secondary leading-[21px] tracking-[0.4px]"
            >
              6 May - 7 May
            </p>
          </div>
        </div>
        <div
          class="tremor-Flex-root flex w-full flex-row justify-start items-start"
          data-testid="total-amount"
        >
          <p
            class="text-tremor-default text-few dark:text-few leading-[22px] font-bold"
          >
            +213%
          </p>
          <p
            class="text-tremor-default ml-1 text-secondary dark:text-dark-romance leading-[21px] tracking-[0.4px]"
          >
            since last month
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="tremor-Card-root w-full text-left rounded-tremor-default bg-tremor-background ring-tremor-ring shadow-tremor-card dark:ring-dark-tremor-ring dark:shadow-dark-tremor-card border-tremor-brand dark:border-dark-tremor-brand dark:bg-dark-tremor-primary ring-0 max-w-full p-4 2xl:max-w-full border-none relative rounded-xl shadow-md"
    >
      <div
        class="tremor-Flex-root flex w-full flex-row justify-between items-start"
      >
        <div
          class="tremor-Flex-root flex justify-between items-center flex-col w-2/3 md:w-1/2"
        >
          <div
            class="tremor-Flex-root flex w-full flex-col justify-start items-start"
            data-testid="formatted-amount"
          >
            <p
              class="text-tremor-default text-md text-secondary dark:text-dark-romance font-semibold tracking-[0.4px]"
            >
              Sales
            </p>
            <p
              class="text-tremor-default text-primary dark:text-dark-primary text-xl leading-[33px] font-bold"
            >
              $23,220.00
            </p>
          </div>
        </div>
        <div
          class="tremor-Flex-root flex flex-row justify-end items-end w-1/3 md:w-1/2 cursor-pointer"
        >
          <p
            class="text-tremor-default !text-xs text-secondary dark:text-secondary leading-[21px] tracking-[0.4px]"
          >
            6 May - 7 May
          </p>
        </div>
      </div>
      <div
        class="tremor-Flex-root flex w-full flex-row justify-start items-start"
        data-testid="total-amount"
      >
        <p
          class="text-tremor-default text-few dark:text-few leading-[22px] font-bold"
        >
          +213%
        </p>
        <p
          class="text-tremor-default ml-1 text-secondary dark:text-dark-romance leading-[21px] tracking-[0.4px]"
        >
          since last month
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
